components/dada-validate/src/validate/validator.rs

```diff
@@ -172,17 +172,46 @@ impl<'me> Validator<'me> {
                 self.add(validated::ExprData::BooleanLiteral(*b), expr)
             }
 
-            syntax::ExprData::IntegerLiteral(w) => {
+            syntax::ExprData::IntegerLiteral(w, suffix_start) => {
                 let raw_str = w.as_str(self.db);
-                let without_underscore: String = raw_str.chars().filter(|&c| c != '_').collect();
-                match u64::from_str(&without_underscore) {
-                    Ok(v) => self.add(validated::ExprData::IntegerLiteral(v), expr),
-                    Err(e) => {
+                let without_underscore: String = raw_str
+                    .chars()
+                    .take(*suffix_start)
+                    .filter(|&c| c != '_')
+                    .collect();
+                let suffix: String = raw_str.chars().skip(*suffix_start).collect();
+                match suffix.as_str() {
+                    "u" => match u64::from_str(&without_underscore) {
+                        Ok(v) => self.add(validated::ExprData::UnsignedIntegerLiteral(v), expr),
+                        Err(e) => {
+                            dada_ir::error!(
+                                self.span(expr),
+                                "`{}` is not a valid integer: {}",
+                                w.as_str(self.db),
+                                e,
+                            )
+                            .emit(self.db);
+                            self.add(validated::ExprData::Error, expr)
+                        }
+                    },
+                    "" | "i" => match i64::from_str(&without_underscore) {
```

Comment:
Hmm. I'm almost tempted to say that unsigned should be the default! But I guess that's nonsense.